sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -3130,42 +3210,358 @@ template <uint32_t SLMSize> __ESIMD_API void slm_init() {
+                                        detail::lsc_data_size::default_size>(
+        byte_offsets, mask, pass_thru);
+  } else {
+    using MsgT = detail::__raw_t<T>;
+    return __esimd_slm_gather_ld<MsgT, N, Alignment>(
+        byte_offsets.data(), mask.data(), pass_thru.data());
+  }
+}
+
+/// template <typename T, int N, int VS,
+///           typename PropertyListT = empty_properties_t>
+/// simd<T, N> slm_gather(simd<uint32_t, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask,
+///                   PropertyListT props = {});                   // (slm_ga-2)
+/// Loads ("gathers") elements of the type 'T' from Shared Local Memory
+/// locations addressed by byte offsets \p byte_offsets, and returns the loaded
+/// elements. Access to any element's memory location can be disabled via the
+/// input vector of predicates \p mask. If mask[i] is unset, then the load from
+/// (byte_offsets[i]) is skipped and the corresponding i-th element of the
+/// returned vector is undefined.
+/// @tparam T Element type.
+/// @tparam N Number of elements to read.
+/// @tparam VS Vector size. It can also be read as the number of reads per each
+/// address. The parameter 'N' must be divisible by 'VS'. (VS > 1) is supported
+/// only on DG2 and PVC.
+/// @param byte_offsets the vector of 32-bit offsets in bytes.
+/// For each i, (byte_offsets[i]) must be element size aligned.
+/// @param mask The access mask, defaults to all 1s.
+/// @param props The optional compile-time properties. Only 'alignment'
+/// property is used.
+/// @return A vector of elements read. Elements in masked out lanes are
+///   undefined.
+template <typename T, int N, int VS,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>, simd<T, N>>
+slm_gather(simd<uint32_t, N / VS> byte_offsets, simd_mask<N / VS> mask,
            PropertyListT props = {}) {
-  using MsgT = detail::__raw_t<T>;
   constexpr size_t Alignment =
       detail::getPropertyValue<PropertyListT, alignment_key>(sizeof(T));
   static_assert(Alignment >= sizeof(T),
                 "slm_gather() requires at least element-size alignment");
-  if constexpr (detail::isMaskedGatherScatterLLVMAvailable()) {
+
+  if constexpr (VS > 1) {
+    return detail::slm_gather_impl<T, VS, detail::lsc_data_size::default_size>(
+        byte_offsets, mask);
+  } else if constexpr (detail::isMaskedGatherScatterLLVMAvailable()) {
+    using MsgT = detail::__raw_t<T>;
     simd<MsgT, N> PassThru; // it is intentionally undefined
     return __esimd_slm_gather_ld<MsgT, N, Alignment>(
         byte_offsets.data(), mask.data(), PassThru.data());
   } else {
-    static_assert(N == 1 || N == 8 || N == 16 || N == 32,
-                  "Unsupported vector length");
+    static_assert(detail::isPowerOf2(N, 32), "Unsupported vector length");
```

Comment:
1) This assert is not needed if you add it to the body of the non-LSC gather_impl.
2) The `if constexpr ` conditions above look not correct. For example, VS=1, N=3, NO llvm.masked.gather allowed case would result into this assert fail.
If not PowerOf2(N,32), and NO llvm.masked.gather, then need to call LSC function.

Reply:
The slm_gather_impl() variant for Gen12 has static_assert() checking that N is powerOf2(N, 32).
You need to check not only VS>1 and if llvm.masked.gather enabled, but also N.
If VS == 1 and N == 3, then slm_gather() for Gen12 must not be called. Please fix the checks in this function to avoid that.

Reply:
Also, the static_assert here is not needed it is already added to slm_gather_impl for Gen12 function.